File: app/src/main/java/net/wuqs/ontime/db/DataTypeConverter.java
package net.wuqs.ontime.db;

import android.arch.persistence.room.TypeConverter;
import android.net.Uri;

import java.util.Calendar;

public class DataTypeConverter {

    @TypeConverter
    public static String toString(Uri uri) {
        return uri.toString();
    }

    @TypeConverter
    public static Uri toUri(String string) {
        return Uri.parse(string);
    }

    public static byte toByte(boolean b) {
        return (byte) toInt(b);
    }

    @TypeConverter
    public static int toInt(boolean b) {
        return b ? 1 : 0;
    }

    @TypeConverter
    public static boolean toBoolean(int n) {
        return n != 0;
    }

    @TypeConverter
    public static Long toLong(Calendar calendar) {
        return calendar.getTimeInMillis();
    }

    @TypeConverter
    public static Calendar toCalendar(Long timeInMillis) {
        Calendar calendar = Calendar.getInstance();
        if (timeInMillis == null) timeInMillis = 0L;
        calendar.setTimeInMillis(timeInMillis);
        return calendar;
    }
}

File: app/src/main/java/net/wuqs/ontime/db/DataTypeConverter.kt
package net.wuqs.ontime.db

import android.arch.persistence.room.TypeConverter
import android.net.Uri
import java.util.*

fun Uri?.toStringOrNull() = this?.toString()

fun String?.toUri() = this?.let { Uri.parse(this) }

fun Boolean.toInt() = if (this) 1 else 0

fun Int.toBoolean() = this != 0

fun Calendar?.toLong() = this?.timeInMillis

fun Long?.toCalendar() = this?.let { Calendar.getInstance().apply { timeInMillis = it } }

class DataTypeConverter {

    @TypeConverter
    fun toStringOrNull(uri: Uri?) = uri?.toString()

    @TypeConverter
    fun toUri(string: String?) = string?.toUri()

    @TypeConverter
    fun toInt(b: Boolean) = b.toInt()

    @TypeConverter
    fun toBoolean(n: Int) = n.toBoolean()

    @TypeConverter
    fun toLong(calendar: Calendar?) = calendar.toLong()

    @TypeConverter
    fun toCalendar(timeInMillis: Long?) = timeInMillis.toCalendar()
}
